designs/filewriter-refactor.md

```diff
@@ -0,0 +1,100 @@
+
+On ownership, the `FileWriter` interface doesn't make clear who is responsible for cleaning up files resulting in ambiguity in consuming code. When considering the `CleanUp()` and `CleanUpTemp()` calls in conjunction with `WithDir()` its difficult to know whether its safe for a top level `FileWriter` to clean up given there's no way to distinguish permanent and temporary files in path contexts created with `WithDir()` contributing further ambiguity to ownership of file lifecycle. When evaluating calls to `CleanUp()` we found it was not called anywhere in the code. With respect to `CleanUpTemp()` we found only a single functional use at the root file writer instance where temporary files are cleaned only if the callpath does not experience an error. Some calls to `CleanUpTemp()` were made immediately after creating a new `FileWriter` instance suggesting the author understood the `NewWriter()` semantics and wanted to immediately delete the `generated` directory `NewWriter()` creates for temporary files.
+
+With respect to the `FileWriter` interface composition, `FileWriter` defines all methods for the concerns outlined earlier directly. This forces consumers to depend on the full `FileWriter` interface even if they do not need to define new path contexts. The vast majority of consumers only care about writing files hence the other 4 methods are superfluous in most contexts.
+
+Finally, the `Write()` method writes files with unix permission bits 0777 before the umask is applied. This brings about risk as files are written with the execution bit for both group and other assuming a default umask of 022.
+
+## Proposal
+
+I propose we reduce the scope of the `FileWriter` to simply providing convinience methods for writing under a path context. Reducing the scope includes removal of the `CleanUp()` and `CleanUpTemp()` methods and any persistent or temporary file options. Additionally, we make clear consumers are responsible for cleaning any files they create with code docs on creation methods. If consumers wish to delete the whole directory they can use the standard libraries `os.RemoveAll(w.Dir())`. Additionally, any files created are assumed to be permanent. Consumers can explicitly define temporary file path contexts that are managed as independent contexts if they need to model temporary files. Finally, I propose a breakdown of interfaces to allow consumers to depend on only what they need and an introduction of a method set that compliments the standard libraries `io` package. Complienting the standard libraries `io` package ensures consumers can better satisfy the SRP as they can function without needing to manage file properties such as name and permissions.
```

Comment:
I suspect it's not an issue, but you made no mention of the potential amount of storage that would be needed to deal with this change? I suspect it's trivial, but thought it might be worth bringing up.

Reply:
Thats a good point. We should probably quantify it so we have the data point either way.